### PR TITLE
[25] Able to add previous education to trainee

### PR DESCRIPTION
**Pull request by asmega**
### Context

- https://trello.com/c/3CF50Pm6/25-add-previous-education-form
- Ability to add previous education data to trainee

### Changes proposed in this pull request

- Adds new page with form for previous education data entry
- Adds and uses faker in factories
- Adds migration for previous education

image

### Guidance to review

- Create or find an existing trainee
- Update their previous education info
- Look in database, values should reflect new values

View https://dfe-twd-rttd-pr-14.herokuapp.com/trainees/1/previous-education

### CODE REVIEW

**Comment by asmega** [comment]
[[M] Add previous education form](https://trello.com/c/3CF50Pm6/25-add-previous-education-form)

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
 <html lang="en" class="govuk-template ">
   <head>
-    <title>GOV.UK Rails Boilerplate</title>
+    <title>Register trainee teachers</title>
```
Comment (defong) [inline app/views/layouts/application.html.erb:4]: Is the repo name wrong or is this or neither?

File: spec/factories/trainees.rb [hunk 2]
```diff
@@ -1,17 +1,51 @@
...
+
+    a_level_1_subject { "Maths" }
+    a_level_1_grade { %w[A B C].sample }
+    a_level_2_subject { "Geography" }
+    a_level_2_grade { %w[A B C].sample }
+    a_level_3_subject { "Physics" }
+    a_level_3_grade { %w[A B C].sample }
+    degree_subject { "Software Engineering" }
+    degree_class { "2:1" }
+    degree_institution { "Durham University" }
```
Comment (defong) [inline spec/factories/trainees.rb:20]: maybe 
https://github.com/faker-ruby/faker/blob/master/doc/default/university.md

File: spec/factories/trainees.rb [hunk 3]
```diff
@@ -1,17 +1,51 @@
 FactoryBot.define do
   factory :trainee do
+    sequence(:trainee_id) { |n| n.to_s }
+    first_names { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    gender { %w[Female Male].sample }
```
Comment (defong) [inline spec/factories/trainees.rb:6]:
```suggestion
    gender { %w[Female Male Other].sample }
```

image

File: spec/factories/trainees.rb [hunk 4]
```diff
@@ -1,17 +1,51 @@
...
+    a_level_3_grade { %w[A B C].sample }
+    degree_subject { "Software Engineering" }
+    degree_class { "2:1" }
+    degree_institution { "Durham University" }
+    degree_type { "BSc" }
+    ske { "SKE not required" }
+    previous_qts { false }
   end
 
   factory :trainee_for_form, class: Trainee do
```
Comment (defong) [inline spec/factories/trainees.rb:26]: What's the diff between this `:trainee_for_form` vs `:trainee`?

Maybe a trait instead?

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (asmega) [reply app/views/layouts/application.html.erb:4]: no idea, i just borrowed this from the prototype as the out of the box one was irking me

File: spec/factories/trainees.rb (hunk 4 above)
Comment (asmega) [reply spec/factories/trainees.rb]: it's do with the date. I've tidied up a little with a base/abstract class type thing. the form take 3 text field whereas the model takes an actual date object

File: spec/factories/trainees.rb (hunk 2 above)
Comment (asmega) [reply spec/factories/trainees.rb]: nice, done

File: spec/factories/trainees.rb (hunk 3 above)
Comment (asmega) [reply spec/factories/trainees.rb]: good spot, done

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (defong) [reply app/views/layouts/application.html.erb:4]: @misaka ^_^

**Review by defong — APPROVED** [review]

**Review by gpeng — COMMENTED** [review]
Can we do a simple feature spec at this point instead of the controller spec? I've gone that route in the course details I'm working on but maybe with the impending design updates it's not worth doing yet. What do you think? I just don't want to establish a pattern of doing controller unit tests that we don't normally do for simple controllers.

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (gpeng) [reply app/views/layouts/application.html.erb:4]: I think that'll probably get settled along the way... probably don't need to worry too much about it at the mo.

Comment (defong) [reply app/views/layouts/application.html.erb:4]: I know... now i seen it also annoys me... :see_no_evil:

**Comment by asmega** [comment]
> Can we do a simple feature spec at this point instead of the controller spec? I've gone that route in the course details I'm working on but maybe with the impending design updates it's not worth doing yet. What do you think? I just don't want to establish a pattern of doing controller unit tests that we don't normally do for simple controllers.

I just tried. I think wait for the index page to be done so we can get find users to update after creating them otherwise we'll always have to make a user first and keep track of them to update.

**Review by gpeng — APPROVED** [review]

**Pull request merged**